File: 0046-permutations/0046-permutations.cpp
class Solution {
public:
    void per(vector<int>&nums,int idx,vector<vector<int>>&ans){
        if(idx>=nums.size()){
            ans.push_back(nums);
            return;
        }
        for(int i=idx;i<nums.size();i++){
            swap(nums[idx],nums[i]);
            per(nums,idx+1,ans);
            swap(nums[idx],nums[i]);
        }
    }
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>>ans;
        int idx=0;
         per(nums,idx,ans);
         return ans;
    }
};

File: 0046-permutations/0046-permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        res=[]
        n=len(nums)
        def back(st):
            if st==n:
                res.append(nums[:])
                return
            for i in range(st,n):
                nums[st],nums[i]=nums[i],nums[st]
                back(st+1)
                nums[st],nums[i]=nums[i],nums[st]
        back(0)
        return res
        